File: DSA/Asmi_Garg/DAY 17/ques1.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    
    TreeNode(int x) {
        val = x;
    }
}

public class BST {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        while (root != null) {
            if (p.val < root.val && q.val < root.val) {
                root = root.left;
            } else if (p.val > root.val && q.val > root.val) {
                root = root.right;
            } else {
                return root;
            }
        }
        return null;
    }

    public static void main(String[] args) {
        BST tree = new BST();
        TreeNode root = new TreeNode(20);
        root.left = new TreeNode(10);
        root.right = new TreeNode(30);
        root.left.left = new TreeNode(5);
        root.left.right = new TreeNode(15);
        root.right.left = new TreeNode(25);
        root.right.right = new TreeNode(35);
        
        TreeNode p = root.left.left; // Node 5
        TreeNode q = root.right;     // Node 30
        
        TreeNode lca = tree.lowestCommonAncestor(root, p, q);
        System.out.println("LCA of " + p.val + " and " + q.val + " is " + lca.val);
    }
}

File: DSA/Vijay-Singh-Chauhan/Day16/ques1.cpp
#include<iostream>
using namespace std;

class Node{
    public:
    int data;
    Node* left;
    Node* right;

    Node(int val){
        this->data=val;
        this->left=NULL;
        this->right=NULL;
    }
};

Node* createnode(){
    int data;
    cout<<"Enter value "<<endl;
    cin>>data;

    if(data==0){
        return NULL;
    }

    Node* root=new Node(data);

    cout<<"left node for "<<root->data<<endl;
    root->left=createnode();

    cout<<"right node for "<<root->data<<endl;
    root->right=createnode();

    return root;
}

void postorder(Node* root){
    if(root==NULL){
        return;
    }

    postorder(root->left);
    postorder(root->right);
    cout<<root->data<<" ";

}
int main(){
    cout<<"Enter 0 for empty node"<<endl;
    Node* root=createnode();

    
    cout<<"post order tree: ";
    postorder(root);
}
